samples/bluetooth/hci_uart/src/main.c

```diff
@@ -52,65 +58,129 @@ static K_FIFO_DEFINE(uart_tx_queue);
  */
 #define H4_DISCARD_LEN 33
 
-static int h4_read(struct device *uart, uint8_t *buf,
-		   size_t len, size_t min)
+static int h4_read(struct device *uart, uint8_t *buf, size_t len)
 {
-	int total = 0;
-
-	while (len) {
-		int rx;
-
-		rx = uart_fifo_read(uart, buf, len);
-		if (rx == 0) {
-			LOG_DBG("Got zero bytes from UART");
-			if (total < min) {
-				continue;
-			}
-			break;
-		}
+	int rx = uart_fifo_read(uart, buf, len);
 
-		LOG_DBG("read %d remaining %d", rx, len - rx);
-		len -= rx;
-		total += rx;
-		buf += rx;
-	}
+	LOG_DBG("read %d req %d", rx, len);
 
-	return total;
+	return rx;
 }
 
-static size_t h4_discard(struct device *uart, size_t len)
+static bool valid_type(uint8_t type)
 {
-	uint8_t buf[H4_DISCARD_LEN];
-
-	return uart_fifo_read(uart, buf, MIN(len, sizeof(buf)));
+	return (type == H4_CMD) | (type == H4_ACL);
 }
 
-static void h4_cmd_recv(struct net_buf *buf, int *remaining)
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static uint32_t get_len(uint8_t *hdr_buf, uint8_t type)
```

Comment:
Nitpicking: it should be `const uint8_t *`. Also those castings below should be to pointers to constant structures.